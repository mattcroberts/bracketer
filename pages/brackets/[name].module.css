.bracketPage {
    display: flex;
    flex-direction: column;
    align-items: center;
}