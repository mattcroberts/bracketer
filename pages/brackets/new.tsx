import { useState } from "react";
import AddBracket from "../../components/forms/AddBracket/AddBracket";
import styles from "./new.module.css";

const NewBracketPage = () => {
  const [noOfBrackets, setNoOfBrackets] = useState(1);
  return (
    <div>
      <h1>New Bracket</h1>

      <form action="/api/create-bracket" method="post" className={styles.form}>
        <label>
          Name <input name="name" />
        </label>

        {new Array(noOfBrackets).fill(0).map((_, index) => (
          <AddBracket key={index} />
        ))}

        <div>
          <button
            type="button"
            onClick={() => {
              setNoOfBrackets(noOfBrackets + 1);
            }}
          >
            Add Team
          </button>
        </div>

        <div>
          <button type="submit">Create Bracket</button>
        </div>
      </form>
    </div>
  );
};

export default NewBracketPage;
