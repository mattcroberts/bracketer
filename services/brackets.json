{
  "123": {
    "name": "123",
    "teams": ["a", "b", "c", "d", "e", "f", "g", "h", "i"],
    "bracket": {
      "stageName": "",
      "a": {
        "stageName": "",
        "a": { "stageName": "", "a": { "team": "a" }, "b": { "team": "b" } },
        "b": { "stageName": "", "a": { "team": "c" }, "b": { "team": "d" } }
      },
      "b": {
        "stageName": "",
        "a": { "stageName": "", "a": { "team": "e" }, "b": { "team": "f" } },
        "b": {
          "stageName": "",
          "a": { "team": "g" },
          "b": { "stageName": "", "a": { "team": "h" }, "b": { "team": "i" } }
        }
      }
    }
  },
  "four": {
    "name": "four",
    "teams": ["A", "B", "C", "D"],
    "bracket": {
      "stageName": "",
      "a": { "stageName": "", "a": { "team": "A" }, "b": { "team": "B" } },
      "b": { "stageName": "", "a": { "team": "C" }, "b": { "team": "D" } }
    }
  },
  "A-H": {
    "name": "A-H",
    "teams": ["A", "B", "C", "D", "E", "F", "G", "H"],
    "bracket": {
      "stageName": "",
      "a": {
        "stageName": "",
        "a": { "stageName": "", "a": { "team": "A" }, "b": { "team": "B" } },
        "b": { "stageName": "", "a": { "team": "C" }, "b": { "team": "D" } }
      },
      "b": {
        "stageName": "",
        "a": { "stageName": "", "a": { "team": "E" }, "b": { "team": "F" } },
        "b": { "stageName": "", "a": { "team": "G" }, "b": { "team": "H" } }
      }
    }
  }
}
