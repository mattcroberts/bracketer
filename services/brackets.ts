import { readFileSync, writeFileSync } from "fs";

export type Bracket = {
  stageName?: string;
  team?: string;
  a?: Bracket;
  b?: Bracket;
};

const BRACKETS: Record<
  string,
  { name: string; teams: Array<string>; bracket: Bracket }
> = {};

export const toPairs = <T>(items: T[]): [T, T][] => {
  const p: [T, T][] = [];

  for (let i = 0; i < items.length; i += 2) {
    p.push([items[i], items[i + 1]]);
  }

  if (items.length % 2) {
    return [
      ...p.slice(0, p.length - 2),
      p[p.length - 2][0],
      [p[p.length - 2][1], p[p.length - 1][0]],
    ];
  }

  return p;
};
const pairPairs = <T>(items: T[]) => {
  if (items.length > 2) {
    return pairPairs(toPairs(items));
  }

  return items;
};

const brackRecur = (paired: any, depth: number): Bracket => {
  const areLeafNodes = paired && paired.length === 1;

  const a = Array.isArray(paired[0])
    ? brackRecur(paired[0], depth + 1)
    : undefined;
  const b = Array.isArray(paired[1])
    ? brackRecur(paired[1], depth + 1)
    : undefined;

  return {
    stageName: "",
    team: undefined,
    a: {
      team: Array.isArray(paired) ? paired[0] : undefined,
      ...a,
    },
    b: {
      team: Array.isArray(paired) ? paired[1] : undefined,
      ...b,
    },
  };
};

export const createBracket = (teams: string[]): Bracket => {
  const pairs = pairPairs(teams);

  return brackRecur(pairs, 0);
};

export const addBracket = (name: string, teams: Array<string>) => {
  const b = JSON.parse(readFileSync("./services/brackets.json").toString());
  b[name] = { name, teams, bracket: createBracket(teams) };
  writeFileSync("./services/brackets.json", JSON.stringify(b));
};

export const getBracket = (name: string) => {
  const b = JSON.parse(readFileSync("./services/brackets.json").toString());

  return b[name];
};
