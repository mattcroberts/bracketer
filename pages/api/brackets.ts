import { NextApiHandler } from "next";
import { FormValues } from "../../components/NewBracketForm";
import { addBracket } from "../../services/brackets";

const createBracket: NextApiHandler = async (req, res) => {
  const teamName: FormValues["name"] = req.body.name;
  const teams: FormValues["teams"] = Array.from(req.body.teams);

  if (teamName && teams.length >= 4) {
    try {
      await addBracket(
        teamName,
        teams.map((team) => team.name as string)
      );

      return res.json({
        teamName,
      });
    } catch (e) {
      console.error(e);
    }
  }
  return res.status(400).json({ error: "Bad Bracket" });
};

export default createBracket;
