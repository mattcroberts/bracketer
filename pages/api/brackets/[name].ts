import { NextApiHandler } from "next";
import { getBracket } from "../../../services/brackets";

const getBracketHandler: NextApiHandler = async (req, res) => {
  const bracketName = req.query.name;

  if (typeof bracketName === "string") {
    try {
      const bracket = await getBracket(bracketName);

      return res.status(200).json(bracket);
    } catch (e) {
      console.error(e);
    }
  }

  return res.status(404).send("Not Found");
};

export default getBracketHandler;
