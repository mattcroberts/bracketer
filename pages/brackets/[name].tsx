import { GetServerSideProps } from "next";

import { Bracket as BracketNode, getBracket } from "../../services/brackets";

import dynamic from "next/dynamic";
import styles from "./[name].module.css";

export const getServerSideProps: GetServerSideProps = async (context) => {
  if (context.params?.name) {
    const bracket = getBracket(context.params.name as string); // TODO

    return { props: { bracket } };
  }

  return { props: {} };
};
const Graph = dynamic(() =>
  import("../../components/BracketGraph/BracketGraph").then(
    (mod) => mod.BracketGraph
  )
);
export default function Bracket({
  bracket,
}: {
  bracket?: { name: string; teams: string[]; bracket: BracketNode };
}) {
  if (!bracket) {
    return null;
  }

  const { name, teams, bracket: bracketNode } = bracket;
  return (
    <div className={styles.bracketPage}>
      <h1>{name} Bracket</h1>

      <h2>Bracket</h2>

      <Graph bracket={bracketNode} />
    </div>
  );
}
