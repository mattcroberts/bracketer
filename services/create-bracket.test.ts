import { createBracket } from "./brackets";

describe("createBracket", () => {
  it.each([
    [["A"]],
    [["A", "B"]],
    [["A", "B", "C"]],
    [["A", "B", "C", "D"]],
    [["A", "B", "C", "D", "E"]],
    [["A", "B", "C", "D", "E", "F"]],
    [["A", "B", "C", "D", "E", "F", "G"]],
    [["A", "B", "C", "D", "E", "F", "G", "H"]],
    [["A", "B", "C", "D", "E", "F", "G", "H", "I"]],
  ])("should create a bracket from %p", (teams) => {
    console.log(teams);

    expect(createBracket(teams)).toMatchSnapshot();
  });
});
