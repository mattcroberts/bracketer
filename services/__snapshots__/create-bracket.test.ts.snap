// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBracket should create a bracket from ["A", "B", "C", "D", "E", "F", "G", "H", "I"] 1`] = `
Object {
  "a": Object {
    "a": Object {
      "a": Object {
        "team": "A",
      },
      "b": Object {
        "team": "B",
      },
      "stageName": "",
      "team": undefined,
    },
    "b": Object {
      "a": Object {
        "team": "C",
      },
      "b": Object {
        "team": "D",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "b": Object {
    "a": Object {
      "a": Object {
        "team": "E",
      },
      "b": Object {
        "team": "F",
      },
      "stageName": "",
      "team": undefined,
    },
    "b": Object {
      "a": Object {
        "team": "G",
      },
      "b": Object {
        "a": Object {
          "team": "H",
        },
        "b": Object {
          "team": "I",
        },
        "stageName": "",
        "team": undefined,
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B", "C", "D", "E", "F", "G", "H"] 1`] = `
Object {
  "a": Object {
    "a": Object {
      "a": Object {
        "team": "A",
      },
      "b": Object {
        "team": "B",
      },
      "stageName": "",
      "team": undefined,
    },
    "b": Object {
      "a": Object {
        "team": "C",
      },
      "b": Object {
        "team": "D",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "b": Object {
    "a": Object {
      "a": Object {
        "team": "E",
      },
      "b": Object {
        "team": "F",
      },
      "stageName": "",
      "team": undefined,
    },
    "b": Object {
      "a": Object {
        "team": "G",
      },
      "b": Object {
        "team": "H",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B", "C", "D", "E", "F", "G"] 1`] = `
Object {
  "a": Object {
    "a": Object {
      "a": Object {
        "team": "A",
      },
      "b": Object {
        "team": "B",
      },
      "stageName": "",
      "team": undefined,
    },
    "b": Object {
      "a": Object {
        "team": "C",
      },
      "b": Object {
        "team": "D",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "b": Object {
    "a": Object {
      "team": "E",
    },
    "b": Object {
      "a": Object {
        "team": "F",
      },
      "b": Object {
        "team": "G",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B", "C", "D", "E", "F"] 1`] = `
Object {
  "a": Object {
    "a": Object {
      "team": "A",
    },
    "b": Object {
      "team": "B",
    },
    "stageName": "",
    "team": undefined,
  },
  "b": Object {
    "a": Object {
      "a": Object {
        "team": "C",
      },
      "b": Object {
        "team": "D",
      },
      "stageName": "",
      "team": undefined,
    },
    "b": Object {
      "a": Object {
        "team": "E",
      },
      "b": Object {
        "team": "F",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B", "C", "D", "E"] 1`] = `
Object {
  "a": Object {
    "a": Object {
      "team": "A",
    },
    "b": Object {
      "team": "B",
    },
    "stageName": "",
    "team": undefined,
  },
  "b": Object {
    "a": Object {
      "team": "C",
    },
    "b": Object {
      "a": Object {
        "team": "D",
      },
      "b": Object {
        "team": "E",
      },
      "stageName": "",
      "team": undefined,
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B", "C", "D"] 1`] = `
Object {
  "a": Object {
    "a": Object {
      "team": "A",
    },
    "b": Object {
      "team": "B",
    },
    "stageName": "",
    "team": undefined,
  },
  "b": Object {
    "a": Object {
      "team": "C",
    },
    "b": Object {
      "team": "D",
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B", "C"] 1`] = `
Object {
  "a": Object {
    "team": "A",
  },
  "b": Object {
    "a": Object {
      "team": "B",
    },
    "b": Object {
      "team": "C",
    },
    "stageName": "",
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A", "B"] 1`] = `
Object {
  "a": Object {
    "team": "A",
  },
  "b": Object {
    "team": "B",
  },
  "stageName": "",
  "team": undefined,
}
`;

exports[`createBracket should create a bracket from ["A"] 1`] = `
Object {
  "a": Object {
    "team": "A",
  },
  "b": Object {
    "team": undefined,
  },
  "stageName": "",
  "team": undefined,
}
`;
